bluemira/optimisation/problem.py

```diff
@@ -0,0 +1,123 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+"""Interface for defining an optimisation problem."""
+import abc
+from typing import Any, Callable, List, Mapping, Optional, Tuple, TypeVar, Union
+
+import numpy as np
+
+from bluemira.optimisation._algorithm import Algorithm
+from bluemira.optimisation._optimise import OptimiserResult, optimise
+from bluemira.optimisation.typing import ConstraintT
+
+
+class OptimisationProblem(abc.ABC):
+    """
+    Interface for an optimisation problem.
+
+    This is an alternative to running an optimisation using the
+    :func:`.optimise` function.
+
+    Using this interface to define an optimisation can provide a few
+    benefits, including:
+
+        * Shared state between optimisation functions and constraints.
+          This can enable things like shared parameters and dynamic
+          constraints.
+        * Switch out optimisation problems using Liskov Substitution.
+        * Logical grouping of related functions.
+    """
+
+    @abc.abstractmethod
+    def objective(self, x: np.ndarray) -> float:
+        """The objective function to minimise."""
+
+    def df_objective(self, x: np.ndarray) -> np.ndarray:
+        """The gradient of the objective function at ``x``."""
+
+    def eq_constraints(self) -> List[ConstraintT]:
+        """The equality constraints on the optimisation."""
+        return []
+
+    def ineq_constraints(self) -> List[ConstraintT]:
+        """The inequality constraints on the optimisation."""
+        return []
+
+    def bounds(self) -> Optional[Tuple[np.ndarray, np.ndarray]]:
+        """The lower and upper bounds of the optimisation parameters."""
+        return None
+
+    def optimise(
+        self,
+        x0: np.ndarray,
+        *,
+        algorithm: Union[Algorithm, str] = Algorithm.SLSQP,
+        opt_conditions: Optional[Mapping[str, Union[int, float]]] = None,
+        opt_parameters: Optional[Mapping[str, Any]] = None,
+        keep_history: bool = False,
+    ) -> OptimiserResult:
+        """
+        Perform the optimisation.
+
+        See :func:`.optimise` for more function parameter details.
+        """
+        return optimise(
+            self.objective,
+            df_objective=self.__overridden_or_default(self.df_objective, None),
+            x0=x0,
+            algorithm=algorithm,
+            opt_conditions=opt_conditions,
+            opt_parameters=opt_parameters,
+            bounds=self.bounds(),
+            eq_constraints=self.eq_constraints(),
+            ineq_constraints=self.ineq_constraints(),
+            keep_history=keep_history,
+        )
+
+    __T1 = TypeVar("__T1", bound=Callable[[Any], Any])
+    __T2 = TypeVar("__T2")
```

Comment:
I don't know how this renders in IDEs, but is it not possible to have slightly more descriptive shorthand here?
Also I don't understand what the TypeVar("__T2") does in practice, but in terms of legibility I am none the wiser :D

Reply:
Yep, better names are a good idea, I'll update them.

`TypeVar` instances basically represent any type (they can also be constrained to specific types, or derivatives of types, by using, e.g., `bound`). So in the case below

```python
T = TypeVar("T")
def f(x: T) -> T:
    ...
```

we're telling the type checker that `f` can take any type as input, and it returns an object of the same type.